rasa/core/policies/rule_policy.py

```diff
@@ -202,12 +190,35 @@ def predict_action_probabilities(
             )
 
         if possible_keys:
+            # TODO rethink that
             key = max(possible_keys, key=len)
 
             recalled = self.lookup.get(key)
+
+            if active_form_name:
+                predicted_listen_from_general_rule = recalled is None or (
+                    domain.action_names[recalled] == ACTION_LISTEN_NAME
+                    and f"active_form_{active_form_name}" not in key
+                )
+                if predicted_listen_from_general_rule:
+                    logger.debug(f"Predicted '{active_form_name}'")
+                    result[domain.index_for_action(active_form_name)] = 1
+                    return result
+
+                predicted_form_from_form_rule = (
+                    domain.action_names[recalled] == active_form_name
+                    and f"active_form_{active_form_name}" in key
+                )
+
+                if predicted_form_from_form_rule:
+                    logger.debug("Added `FormValidation(False)` event.")
+                    tracker.update(FormValidation(False))
+
             if recalled is not None:
+
                 logger.debug(
-                    f"There is a memorised next action '{domain.action_names[recalled]}'"
+                    f"There is a memorised next action "
+                    f"'{domain.action_names[recalled]}'"
                 )
 
                 if self.USE_NLU_CONFIDENCE_AS_SCORE:
```

Comment:
the code after line 240 seems weird. Can we remove it?

Reply:
This is a copy paste form memo. I think we can remove it or substitute with it everywhere